### PR TITLE
Copy color to depth, defer depth buffer copies

### LINKED ISSUE #15728: Better ways to deal with overlapping render targets
In #15717, I fixed the Juiced 2 rendering issue where two render targets share a bit of data by partially overlapping, by allowing PPSSPP to consider them the same render target. Since a lot of draws are done texturing from one and writing to the other, this leads to a lot of detected "self-texturing" which leads to a large number of expensive copies.

An alternative approach would be to leave the two render targets separate, but on bind, detect other overlapping rendertargets and copy the overlapping areas from the other ones if they're "fresher". A bit similar to what #15700 will do for depth buffers, but with an offset. That would avoid the self-texturing issue entirely, but adds a little bit of complexity that we might not always want to have enabled.

Anyway, if implemented, this should get Juiced 2 speed back to at least where it was before, so might be worthwhile in the future.

### PR DESCRIPTION
**Pull request by hrydgard**
~~Goes on top of #15857 (currently includes all of it, but will be rebased once that's merged).~~

This supersedes #15777 , and includes all of it. Anyway, it:

* Removes the color-to-depth drawing mode, in favor of color-to-depth copies on bind
* Makes the depth copy flow more logical, and a template for the upcoming color copy flow that will take care of #15728
* In case of missing depth texture support, binds a null texture instead of going to RAM.
* Cleans up and refactors a bit

This has one known regression, which we can live with temporarily (the fix might shake out of future changes):

- [ ] Colin McRae shadow flicker returns (#7810)

### CODE REVIEW

**Comment by hrydgard** [comment]
I was hoping that this would fix #15491 (ratchet & clank) but the stumbling block there is how it uses a raw 565 copy (sampling as 565, writing 565) to stretch the depth buffer. Our matching logic rejects that.

It'll require something I've planned as one of the next steps, which is that if we texture-sample the "wrong" format, we create a clone of the framebuffer at the same address but with the desired format, and then let the planned logic to copy to color from overlapping buffers handle it (call CopyToColorFromOverlappingFramebuffers on it before texturing from it). Alternatively, it could be handled like depal, but without the CLUT. Might actually be the most straightforward path, though least generic. Not sure...

**Comment by hrydgard** [comment]
I'm gonna get this in soon, it seems to work well now, with the lone exception of Colin McRae. I do plan to fix that one again, when I go look at the other problems it has (like its sun effect).

File: GPU/Common/FramebufferManagerCommon.cpp [hunk 1]
```diff
@@ -538,28 +528,36 @@ void FramebufferManagerCommon::CopyToDepthFromOverlappingFramebuffers(VirtualFra
...
 		if (source.channel == RASTER_DEPTH) {
 			// Good old depth->depth copy.
 			BlitFramebufferDepth(source.vfb, dest);
-
 			gpuStats.numDepthCopies++;
 			dest->last_frame_depth_updated = gpuStats.numFlips;
+		} else if (source.channel == RASTER_COLOR) {
+			VirtualFramebuffer *src = source.vfb;
+			// Copying color to depth.
+			BlitUsingRaster(src->fbo, 0.0f, 0.0f, src->renderWidth, src->renderHeight, dest->fbo, 0.0f, 0.0f, dest->renderWidth, dest->renderHeight, false, DRAW2D_565_TO_DEPTH, "565_to_depth");
```
Comment (unknownbrackets) [inline GPU/Common/FramebufferManagerCommon.cpp:565]: nit: Maybe log drawnFormat if not 565?  Maybe related to the Colin thing?

-[Unknown]

File: GPU/Common/FramebufferManagerCommon.cpp [hunk 2]
```diff
@@ -523,16 +524,17 @@ void FramebufferManagerCommon::CopyToDepthFromOverlappingFramebuffers(VirtualFra
...
 			continue;
 
 		if (src->fb_address == dest->z_address && src->fb_stride == dest->z_stride && src->format == GE_FORMAT_565) {
-			if (src->colorBindSeq < dest->depthBindSeq) {
+			if (src->colorBindSeq > dest->depthBindSeq) {
 				// Source has newer data than the current buffer, use it.
 				sources.push_back(CopySource{ src, RASTER_COLOR });
 			}
 		} else if (src->z_address == dest->z_address && src->z_stride == dest->z_stride && src->depthBindSeq > dest->depthBindSeq) {
+			// Don't bother if the buffer was from another frame. This heuristic is old.
```
Comment (unknownbrackets) [inline GPU/Common/FramebufferManagerCommon.cpp:540]: Should we not bother?  Although most games re-render depth every frame, it wouldn't shock me for a game to keep it around for render-to-texture in a dialog scene (i.e. a frozen image behind, blurred using depth.)  Not sure if any do that, though.

But I think this comment claims something that isn't currently done anyway.

-[Unknown]

File: GPU/Common/FramebufferManagerCommon.cpp [hunk 3]
```diff
@@ -1957,7 +1959,7 @@ void FramebufferManagerCommon::UpdateFramebufUsage(VirtualFramebuffer *vfb) {
 
 	checkFlag(FB_USAGE_DISPLAYED_FRAMEBUFFER, vfb->last_frame_displayed);
 	checkFlag(FB_USAGE_TEXTURE, vfb->last_frame_used);
-	checkFlag(FB_USAGE_RENDERTARGET, vfb->last_frame_render);
+	checkFlag(FB_USAGE_RENDER_COLOR, vfb->last_frame_render);
```
Comment (unknownbrackets) [inline GPU/Common/FramebufferManagerCommon.cpp:1965]: Hm, what if it's only rendering depth?  I could see it being useful to know that... hm.  Maybe for render-to-texture selection too.  Can't remember if I've seen games render a scene with RGB and stencil all masked out, just to establish depth... feel like I have...

-[Unknown]

File: GPU/Common/FramebufferManagerCommon.cpp (hunk 2 above)
Comment (hrydgard) [reply GPU/Common/FramebufferManagerCommon.cpp:540]: Hm that's right, this comment was wrong and based on a misunderstanding I had for a while. I'll clean it up.

File: GPU/Common/FramebufferManagerCommon.cpp (hunk 3 above)
Comment (hrydgard) [reply GPU/Common/FramebufferManagerCommon.cpp:1965]: Yeah, this tracking isn't perfect yet. I don't think I have seen such a case though, but I also don't think it'll hurt very much if we get that wrong..

File: GPU/Common/FramebufferManagerCommon.cpp (hunk 1 above)
Comment (hrydgard) [reply GPU/Common/FramebufferManagerCommon.cpp:565]: I'm planning to remove drawnFormat soon, and instead have duplicate VirtualFramebuffer, and automatic reinterpret copies between them on-demand. But yeah, can certainly add such logging in the meantime.

File: GPU/GPUCommon.cpp [hunk 4]
```diff
@@ -1685,6 +1685,16 @@ void GPUCommon::Execute_Prim(u32 op, u32 diff) {
 		return;
 	}
 
+	if (!gstate_c.usingDepth) {
+		bool isClearingDepth = gstate.isModeClear() && gstate.isClearModeDepthMask();;
```
[the remark this answers is not in the document]
Comment (hrydgard) [reply GPU/GPUCommon.cpp]: this was later removed.

**Pull request merged**